quil-rs/src/instruction/gate.rs

```diff
@@ -128,13 +137,77 @@ impl fmt::Display for GateModifier {
     }
 }
 
+#[derive(Copy, Clone, Debug, PartialEq, Eq, strum::Display, strum::EnumString)]
+#[strum(serialize_all = "UPPERCASE")]
+pub enum PauliGate {
+    I,
+    X,
+    Y,
+    Z,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PauliTerm {
+    pub word: Vec<PauliGate>,
+    pub expression: Expression,
+    pub arguments: Vec<String>,
+}
+
+impl PauliTerm {
+    pub fn new(
+        word: Vec<PauliGate>,
+        expression: Expression,
+        arguments: Vec<String>,
+    ) -> Result<Self, GateError> {
+        if word.len() != arguments.len() {
+            Err(GateError::PauliTermArgumentLength {
+                expected: word.len(),
+                actual: arguments.len(),
+            })
+        } else {
+            Ok(Self {
+                word,
+                expression,
+                arguments,
+            })
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PauliSum {
+    pub arguments: Vec<String>,
+    pub terms: Vec<PauliTerm>,
+}
+
+impl PauliSum {
+    pub fn new(arguments: Vec<String>, terms: Vec<PauliTerm>) -> Result<Self, GateError> {
+        let term_arguments =
+            HashSet::<String>::from_iter(terms.iter().flat_map(|t| t.arguments.clone()));
+        let argument_set = HashSet::from_iter(arguments.clone().into_iter());
+        let diff: Vec<&String> = term_arguments.difference(&argument_set).collect();
```

Comment:
Anyone have the Rust chops to do this without all the cloning? I tried fighting with the borrow checker for a while with no luck. It feels like there should be a better way to do this, but it's not coming to me.

Reply:
This seems to compile but not sure if it is better 

```rs
        let term_arguments = terms
            .iter()
            .flat_map(|t| t.arguments.iter())
            .collect::<HashSet<_>>();

        let argument_set = arguments.iter().collect::<HashSet<_>>();

        let diff = term_arguments.difference(&argument_set).collect::<Vec<_>>();

        if !diff.is_empty() {
            return Err(GateError::PauliSumArgumentMismatch {
                expected_arguments: arguments.clone(),
                mismatches: diff.into_iter().map(|&v| v.clone()).collect(),
            });
        }
```

Reply:
Or this? /shrug

```rs
        let diff = terms
            .iter()
            .flat_map(|t| t.arguments.iter())
            .collect::<HashSet<_>>()
            .difference(&arguments.iter().collect::<HashSet<_>>())
            .map(|v| v.to_owned().to_owned()) // probably a better way
            .collect::<Vec<_>>();

        if !diff.is_empty() {
            return Err(GateError::PauliSumArgumentMismatch {
                expected_arguments: arguments,
                mismatches: diff,
            });
        }

        Ok(Self { arguments, terms })
```

Reply:
In @jselig-rigetti's second example, I think you can replace `.map(|v| v.to_owned().to_owned())` with `.copied().cloned()` -- references are `Copy`, so `copied` should dereference the second reference and `cloned` should convert the leftover references into owned values.

A second option is `.map(|&v| v.to_owned())`, which implicitly dereferences the outer reference via pattern matching.

Reply:
Cloned this locally and tweaked @jselig-rigetti's suggestion a little to get this, which at least passes `cargo check`:

```rust
        let diff = terms
            .iter()
            .flat_map(|t| t.arguments.iter().map(|(_, a)| a))
            .collect::<HashSet<_>>()
            .difference(&arguments.iter().collect::<HashSet<_>>())
            .copied()
            .cloned()
            .collect::<Vec<_>>();

        if !diff.is_empty() {
            return Err(GateError::PauliSumArgumentMismatch {
                expected_arguments: arguments,
                mismatches: diff,
            });
        }

        Ok(Self { arguments, terms })

```

Reply:
Nice! Thanks for playing golf! It seems like collecting into a `HashSet` was the way to go after all. I hit a dead end with that approach the first try, but this last suggestion works perfectly. I made a small tweak so I could shove the `clone()` into the error case.